runtime/fs_node.js

```diff
@@ -104,16 +106,136 @@ MlNodeDevice.prototype.open = function(name, f) {
     var fd = this.fs.openSync(this.nm(name), res);
     return new MlNodeFile(fd);
   } catch (err) {
-    caml_raise_sys_error(err.toString());
+    this.raise_unix_exn_of_nodejs_error(err);
   }
 }
 
 MlNodeDevice.prototype.rename = function(o,n) {
   try {
     this.fs.renameSync(this.nm(o), this.nm(n));
   } catch (err) {
-    caml_raise_sys_error(err.toString());
+    this.raise_unix_exn_of_nodejs_error(err);
+  }
+}
+MlNodeDevice.prototype.stat = function(name) {
+  try {
+    var js_stats = this.fs.statSync(this.nm(name));
+    return this.stats_from_js(js_stats);
+  } catch (err) {
+    this.raise_unix_exn_of_nodejs_error(err);
+  }
+}
+MlNodeDevice.prototype.lstat = function(name) {
+  try {
+    var js_stats = this.fs.lstatSync(this.nm(name));
+    return this.stats_from_js(js_stats);
+  } catch (err) {
+    this.raise_unix_exn_of_nodejs_error(err);
+  }
+}
+MlNodeDevice.prototype.symlink = function(to_dir, src, dst) {
+  try {
+    this.fs.symlinkSync(this.nm(src), this.nm(dst), to_dir ? 'dir' : 'file');
+    return 0;
+  } catch (err) {
+    this.raise_unix_exn_of_nodejs_error(err);
+  }
+}
+MlNodeDevice.prototype.readlink = function(name) {
+  try {
+    var js_stats = this.fs.readlinkSync(this.nm(name), 'utf8');
+    return caml_string_of_jsbytes(js_stats);
+  } catch (err) {
+    this.raise_unix_exn_of_nodejs_error(err);
+  }
+}
+MlNodeDevice.prototype.raise_unix_exn_of_nodejs_error = function(err) {
+  /* ===Unix.error===
+   *
+   * This array is in order of the variant in OCaml
+   */
+  var errors = [
+    "E2BIG", "EACCES", "EAGAIN", "EBADF", "EBUSY", "ECHILD", "EDEADLK", "EDOM",
+    "EEXIST", "EFAULT", "EFBIG", "EINTR", "EINVAL", "EIO", "EISDIR", "EMFILE",
+    "EMLINK", "ENAMETOOLONG", "ENFILE", "ENODEV", "ENOENT", "ENOEXEC", "ENOLCK",
+    "ENOMEM", "ENOSPC", "ENOSYS", "ENOTDIR", "ENOTEMPTY", "ENOTTY", "ENXIO",
+    "EPERM", "EPIPE", "ERANGE", "EROFS", "ESPIPE", "ESRCH", "EXDEV", "EWOULDBLOCK",
+    "EINPROGRESS", "EALREADY", "ENOTSOCK", "EDESTADDRREQ", "EMSGSIZE",
+    "EPROTOTYPE", "ENOPROTOOPT", "EPROTONOSUPPORT", "ESOCKTNOSUPPORT",
+    "EOPNOTSUPP", "EPFNOSUPPORT", "EAFNOSUPPORT", "EADDRINUSE", "EADDRNOTAVAIL",
+    "ENETDOWN", "ENETUNREACH", "ENETRESET", "ECONNABORTED", "ECONNRESET", "ENOBUFS",
+    "EISCONN", "ENOTCONN", "ESHUTDOWN", "ETOOMANYREFS", "ETIMEDOUT", "ECONNREFUSED",
+    "EHOSTDOWN", "EHOSTUNREACH", "ELOOP", "EOVERFLOW"
+  ];
+  var variant = errors.indexOf(err.code);
+  if (variant < 0) {
+    variant = BLOCK(0, err.errno);
```

Comment:
A comment explaining this is EUNKNOWNERR would be nice. What if errno is not defined ?

Reply:
I believe errno is always supposed to be attached to a node "fs" error. Do you want me to add a fallback int?

Reply:
Comment added, let me know if there's a fallback in that makes sense 👍